### ISSUE TITLE
Ramses applications freeze with AMD Radeon Windows graphics drivers

### ISSUE THREAD

**Issue by AchimStremplat** [bmwcarit/ramses#61]
Running the ramses renderer as a stand-alone executable, or as part of the [ramses-logic-viewer](https://ramses-logic.readthedocs.io/en/latest/viewer.html) or as part of the [Ramses Composer](https://github.com/COVESA/ramses-composer) does not work on a system using Windows 10 and an AMD Radeon chipset (I was able to reproduce it on a system with an AMD Radeon RX6600 using the Radeon Graphics Drivers Adrenalin Edition 22.3.2 - but also with the default Microsoft Windows drivers).

The application freezes shortly after startup in one of the first few SwapBuffers calls in Context_WGL::swapBuffers().

After some experimentation I found that the problem vanishes if a "glFlush()" is added before the "return true;" in ShaderUploader_GL::UploadShaderProgramFromSource(...) before line 111 (as of commit [a452f8c](https://github.com/COVESA/ramses/commit/a452f8cd8cef6203c2c5fc9a733edc08bdc8446d)).

I can't say that I really understand why (or if the change is correct), except that the [documentation for SwapBuffers](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-swapbuffers) says "With multithread applications, flush the drawing commands in any other threads drawing to the same window before calling SwapBuffers."

The problem is specific to AMD GPUs, it does not occur on NVidia GPUs/drivers or Intel GPUs/drivers.

If you need me to test any changes to Ramses on my system, just let me know.

**Comment by violinyanev** [comment]
Thanks for the report @AchimStremplat! Looks like quite the entangled bug, but we can work with this.

My working hypothesis - let's try to prove/disprove it:
- The issue is in the Windows-specific implementation of egl (wgl), which potentially has a different behavior for different GPU families - I could not find the same "suggestion" on EGL, other than it implicitly flushes on the _same_ thread where eglSwapBuffers is called
- The crash results from a data race when using multiple threads to communicate with the GPU (which ramses does since v27.0.105). Ramses only uses the extra thread to upload shaders to the GPU - which would explain why adding glFlush there hides the issue (the "other thread" is the one which uploads shaders)
- The crash happens on the Kernel-side of the driver (which means there is probably a real bug in the Windows drivers - it should not be possible to kill the driver by incorrect thread handling)
- I can imagine Windows might be interested in the bug - it is a potential security threat. Will make sure to report it on MSDN


Can you perform one or more of the following tests:
1. Check if a non-Windows application also crashes (I think a Linux VM won't be enough because it virtualizes the GPU, but if you can e.g. install Ubuntu from a live CD/USB and test if it the issue happens there, we could prove or disprove its Windows only, given the same GPU
2. Check if a version of Ramses without threads has the same issue. From what I can see, the 'ramses-example-local-datalink-xxx' example is implemented without threads, i.e. everything is called on the same thread. If we don't have the issue there, it would confirm that the cause is truly the missing glFlush on the resource-uploading thread.

Thanks a lot for the help!

**Comment by ParadoxKG** [comment]
I can confirm that the suggested fix also works on an AMD RX 580 and Windows 10.
Original Ramses crashes the display driver (black screen for 2 seconds while it re-initializes).
Fixed version runs fine.

**Comment by violinyanev** [comment]
@ParadoxKG  can you run the ramses-example-local-datalink-xxx example without the fix to verify it doesn't crash there?

**Comment by mohamedsharaf-el-deen-bmw-ag** [comment]
Hello,
Thanks @AchimStremplat  and @ParadoxKG for the detailed analysis.

As @violinyanev  mentioned, we do not render in the 2nd context we only use it for shader upload, so theoretically we do not need to follow Microsoft doc's guideline about adding a flush. I think we paid special attention to that point while adding the feature and decided it was unnecessary. However, the situation we always have to face is possible difference in behavior of GL/EGL implementations from one vendor to another. The GL/EGL specs do not make the situation easier since they are famous for being under-specified, which only leaves place for more variation in interpretation and variation in behavior between the different vendors. We try to account for that by running our tests on different platforms, but of course we can only cover a handful of combinations, so here we are! :) 

I tried to find if there were any bug reports to AMD with the keywords that relate to the situation here but I could not find a 100% match. 

Nevertheless, I could find some discussions/reports that confirm the possibility that AMD driver(s) might require a glFlush in order to sync objects from one context to another. In my (limited) experience, that's usually the nearest you can get to a proof that the problem exists without filing a bug (to AMD) yourself/ourselves and demanding investigation of the very specific issue.

https://stackoverflow.com/questions/33640702/opengl-glclientwaitsync-on-separate-thread
https://community.amd.com/t5/archives-discussions/opengl-opencl-interop-with-shared-contexes-and-multithreading/td-p/262346

To emphasize, I think the current evidence points to a difference in behavior in AMD drivers (at least on Windows, i.e., WGL?) from what was always witnessed on other platform available at hand. I agree that we might need to add the glFlush to the shader-upload-context. If this does not lead to a very apparent problem with other vendors' implementations then it should be the way to go in my opinion as well.

Note to self: special attention to the change needs to be taken, since this change might result in an unwanted performance overhead on some target platforms especially some that are used in production.

**Comment by ParadoxKG** [comment]
I tested with the ramses-example-local-datalink-xxx example with AMD RX 580. The behaviour is exactly the same: Crash/blackscreen in original configuration, running nicely with fixed DLL.

**Comment by AchimStremplat** [comment]
Thanks for the quick reply and in-depth explanation. 

> I tested with the ramses-example-local-datalink-xxx example with AMD RX 580. The behaviour is exactly the same: Crash/blackscreen in original configuration, running nicely with fixed DLL.

I can confirm that. I double-checked and the process calls Context_WGL::initCustomPixelFormat twice (just like the other applications do) - once from the main thread, and once from AsyncEffectUploader::run(). The second thread also seems to do stuff, i. e. it calls AsyncEffectUploader::uploadEffectsOrWait which does call resourceUploadRenderBackend.getDevice().uploadShader (tried it with ramses-example-local-datalink-windows-wgl-4-5). 

Here is the callstack which creates the second thread (in case that helps):

```
>	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::AsyncEffectUploader::createResourceUploadRenderBackendAndStartThread() Line 44	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::RendererSceneUpdater::createDisplayContext(const ramses_internal::DisplayConfig & displayConfig, ramses_internal::IBinaryShaderCache * binaryShaderCache) Line 122	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::RendererCommandExecutor::operator()(const ramses_internal::RendererCommand::CreateDisplay & cmd) Line 128	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::base_internal::Callable::Invoke<ramses_internal::RendererCommandExecutor &,ramses_internal::RendererCommand::CreateDisplay &>(ramses_internal::RendererCommandExecutor & f, ramses_internal::RendererCommand::CreateDisplay & <args_0>) Line 154	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::base_internal::Invoke<ramses_internal::RendererCommandExecutor &,ramses_internal::RendererCommand::CreateDisplay &>(ramses_internal::RendererCommandExecutor & f, ramses_internal::RendererCommand::CreateDisplay & <args_0>) Line 182	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &>::Run<0,12>(std::integral_constant<bool,0> __formal, absl::lts_2019_08_08::integer_sequence<unsigned __int64,0> __formal, std::integral_constant<unsigned __int64,12> __formal) Line 937	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &>::operator()<12>(std::integral_constant<unsigned __int64,12> <indices_0>) Line 923	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::variant_internal::call_with_indices<void,absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &>,12>(absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &> && function) Line 218	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::variant_internal::VisitIndicesFallback<44>::Run<absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &>,unsigned __int64>(absl::lts_2019_08_08::variant_internal::PerformVisitation<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &> && op, unsigned __int64 <indices_0>) Line 441	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!absl::lts_2019_08_08::visit<ramses_internal::RendererCommandExecutor &,absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> &>(ramses_internal::RendererCommandExecutor & vis, absl::lts_2019_08_08::variant<ramses_internal::RendererCommand::ScenePublished,ramses_internal::RendererCommand::SceneUnpublished,ramses_internal::RendererCommand::ReceiveScene,ramses_internal::RendererCommand::UpdateScene,ramses_internal::RendererCommand::SetSceneState,ramses_internal::RendererCommand::SetSceneMapping,ramses_internal::RendererCommand::SetSceneDisplayBufferAssignment,ramses_internal::RendererCommand::LinkData,ramses_internal::RendererCommand::LinkOffscreenBuffer,ramses_internal::RendererCommand::LinkStreamBuffer,ramses_internal::RendererCommand::UnlinkData,ramses_internal::RendererCommand::PickEvent,ramses_internal::RendererCommand::CreateDisplay,ramses_internal::RendererCommand::DestroyDisplay,ramses_internal::RendererCommand::CreateOffscreenBuffer,ramses_internal::RendererCommand::CreateDmaOffscreenBuffer,ramses_internal::RendererCommand::DestroyOffscreenBuffer,ramses_internal::RendererCommand::CreateStreamBuffer,ramses_internal::RendererCommand::DestroyStreamBuffer,ramses_internal::RendererCommand::SetStreamBufferState,ramses_internal::RendererCommand::SetClearFlags,ramses_internal::RendererCommand::SetClearColor,ramses_internal::RendererCommand::UpdateWarpingData,ramses_internal::RendererCommand::ReadPixels,ramses_internal::RendererCommand::SetSkippingOfUnmodifiedBuffers,ramses_internal::RendererCommand::LogStatistics,ramses_internal::RendererCommand::LogInfo,ramses_internal::RendererCommand::SCListIviSurfaces,ramses_internal::RendererCommand::SCSetIviSurfaceVisibility,ramses_internal::RendererCommand::SCSetIviSurfaceOpacity,ramses_internal::RendererCommand::SCSetIviSurfaceDestRectangle,ramses_internal::RendererCommand::SCScreenshot,ramses_internal::RendererCommand::SCAddIviSurfaceToIviLayer,ramses_internal::RendererCommand::SCSetIviLayerVisibility,ramses_internal::RendererCommand::SCRemoveIviSurfaceFromIviLayer,ramses_internal::RendererCommand::SCDestroyIviSurface,ramses_internal::RendererCommand::SetLimits_FrameBudgets,ramses_internal::RendererCommand::SetLimits_FlushesForceApply,ramses_internal::RendererCommand::SetLimits_FlushesForceUnsubscribe,ramses_internal::RendererCommand::FrameProfiler_Toggle,ramses_internal::RendererCommand::FrameProfiler_TimingGraphHeight,ramses_internal::RendererCommand::FrameProfiler_CounterGraphHeight,ramses_internal::RendererCommand::FrameProfiler_RegionFilterFlags,ramses_internal::RendererCommand::ConfirmationEcho> & <vars_0>) Line 438	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::RendererCommandExecutor::executePendingCommands() Line 49	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::DisplayBundle::update() Line 71	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::DisplayBundle::doOneLoop(ramses_internal::ELoopMode loopMode, std::chrono::duration<__int64,std::ratio<1,1000000>> sleepTime) Line 54	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses_internal::DisplayDispatcher::doOneLoop(std::chrono::duration<__int64,std::ratio<1,1000000>> sleepTime) Line 89	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses::RamsesRendererImpl::doOneLoop() Line 93	C++
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses::RamsesRenderer::doOneLoop() Line 29	C++
 	ramses-example-local-datalink-windows-wgl-4-5.exe!SceneStateEventHandler::waitForDataProviderOrConsumerCreated(ramses::StronglyTypedValue<unsigned __int64,0,ramses::SceneIdTag> sceneId) Line 71	C++
 	ramses-example-local-datalink-windows-wgl-4-5.exe!main(int argc, char * * argv) Line 348	C++
```

The RenderedCommand::CreateDisplay which seems(?) to be responsible for causing the callstack might be created by
```
 	ramses-shared-lib-windows-wgl-4-5.dll!ramses::RamsesRendererImpl::createDisplay(const ramses::DisplayConfig & config) Line 145	C++
>	ramses-shared-lib-windows-wgl-4-5.dll!ramses::RamsesRenderer::createDisplay(const ramses::DisplayConfig & config) Line 87	C++
 	ramses-example-local-datalink-windows-wgl-4-5.exe!main(int argc, char * * argv) Line 275	C++
```
though its possible that I get things confused here.

In short: it looks like the sample also makes use of two threads, so it is not surprising that it crashes on systems with AMD drivers without the glFlush.

>Check if a non-Windows application also crashes (I think a Linux VM won't be enough because it virtualizes the GPU, but if you can e.g. install Ubuntu from a live CD/USB and test if it the issue happens there, we could prove or disprove its Windows only, given the same GPU

I tried that using Ubuntu 20.04 with the open source AMD driver - in that environment RaCo can be started without issues using the unchanged Ramses (no "glFlush") even with an AMD Radeon RX6600 installed.

**Comment by violinyanev** [comment]
Thanks @AchimStremplat ! We are working on a fix (or rather workaround) based on your findings. I looked for ways to open a windows bug, but it doesn't seem to be possible (I was really surprised)...
I tried second to file a bug to the AMD driver using [this](https://www.amd.com/en/support/kb/faq/amdbrt?utm_campaign=report&utm_medium=redirect&utm_source=301), but the tool refuses to report a bug because I don't have an AMD GPU... The process looks quite simple, let's try and submit the issue there?

**Comment by violinyanev** [comment]
Should work with v27.0.119. Can you verify @AchimStremplat ?

**Comment by AchimStremplat** [comment]
Works. Thanks!
